include/maths.h

```diff
@@ -45,23 +45,46 @@ inline uint64_t inv64_u64(uint64_t p) {
 }
 
 // montgomery multiply
-inline uint64_t mont_mul(uint64_t a, uint64_t b, uint64_t p, uint64_t p_dash) {
-  uint128_t t = (uint128_t)a * b;
-  uint64_t m = (uint64_t)t * p_dash;
-  uint128_t u = t + (uint128_t)m * p;
-  uint64_t res = u >> 64;
-  int64_t maybe = res - p;
-  return maybe < 0 ? res : (uint64_t)maybe;
+// static inline uint64_t mont_mul(uint64_t a, uint64_t b, uint64_t p, uint64_t p_dash) {
+//   uint128_t t = (uint128_t)a * b;
+//   uint64_t m = (uint64_t)t * p_dash;
+//   uint128_t u = t + (uint128_t)m * p;
+//   uint64_t res = u >> 64;
+//   int64_t maybe = res - p;
+//   return maybe < 0 ? res : (uint64_t)maybe;
+// }
+
+static inline uint64_t mont_mul(uint64_t a, uint64_t b, uint64_t p, uint64_t p_dash) {
+  uint64_t  result;
+  uint64_t scratch;
+    asm (
+        ".intel_syntax\n"
+        "mul     %[b]\n"
+        "mov     %[working], %%rdx\n"
+        "imul    %%rax, %[p_dash]\n"
+        "mul     %[p]\n"
+        "add     %%rax, -1\n"
+        "adc     %%rdx, %[working]\n"
+        "mov     %%rax, %%rdx\n"
+        "sub     %%rax, %[p]\n"
+        "cmovs   %%rax, %%rdx\n"
+        ".att_syntax\n"
+        : "=a"(result)
```

Comment:
Apparently the "a" constraint will make this `rax`... but with some hand waving about "well yes because you're compiling for 64 bit" because "a" also refers to `eax` - this makes me uncomfortable.